test-suite/js/src/test/scala/org/scalajs/testsuite/javalib/util/ArrayDequeTest.scala

```diff
@@ -0,0 +1,193 @@
+      expect(llString.add("pippo")).toBeTruthy
+      expect(llString.pollLast()).toEqual("pippo")
+
+      val llDouble = new ArrayDeque[Double]()
+
+      expect(llDouble.add(+10000.987)).toBeTruthy
+      expect(llDouble.add(-0.987)).toBeTruthy
+      expect(llDouble.pollLast()).toEqual(-0.987)
+    }
+
+    it("should perform as a stack with push and pop") {
+      val llInt = new ArrayDeque[Int]()
+
+      llInt.push(1000)
+      llInt.push(10)
+      expect(llInt.pop()).toEqual(10)
+      expect(llInt.pop()).toEqual(1000)
+      expect(llInt.isEmpty()).toBeTruthy
+
+      val llString = new ArrayDeque[String]()
+
+      llString.push("pluto")
+      llString.push("pippo")
+      expect(llString.pop()).toEqual("pippo")
+      expect(llString.pop()).toEqual("pluto")
+      expect(llString.isEmpty()).toBeTruthy
+
+      val llDouble = new ArrayDeque[Double]()
+
+      llDouble.push(+10000.987)
+      llDouble.push(-0.987)
+      expect(llDouble.pop()).toEqual(-0.987)
+      expect(llDouble.pop()).toEqual(+10000.987)
+      expect(llString.isEmpty()).toBeTruthy
+    }
+
+    it("should poll and peek elements") {
+      val pq = new ArrayDeque[String]()
+
+      expect(pq.add("one")).toBeTruthy
+      expect(pq.add("two")).toBeTruthy
+      expect(pq.add("three")).toBeTruthy
+
+      expect(pq.peek.equals("one")).toBeTruthy
+      expect(pq.poll.equals("one")).toBeTruthy
+
+      expect(pq.peekFirst.equals("two")).toBeTruthy
+      expect(pq.pollFirst.equals("two")).toBeTruthy
+
+      expect(pq.peekLast.equals("three")).toBeTruthy
+      expect(pq.pollLast.equals("three")).toBeTruthy
+
+      expect(pq.peekFirst).toBeNull
+      expect(pq.pollFirst).toBeNull
+
+      expect(pq.peekLast).toBeNull
+      expect(pq.pollLast).toBeNull
+    }
+
+    it("should remove occurrences of provided elements") {
+      val l = asJavaCollection(Seq("one", "two", "three", "two", "one"))
+      val ll = new ArrayDeque[String](l)
+
+      expect(ll.removeFirstOccurrence("one")).toBeTruthy
+      expect(ll.removeLastOccurrence("two")).toBeTruthy
+      expect(ll.removeFirstOccurrence("one")).toBeTruthy
+      expect(ll.removeLastOccurrence("two")).toBeTruthy
+      expect(ll.removeFirstOccurrence("three")).toBeTruthy
+      expect(ll.removeLastOccurrence("three")).toBeFalsy
+      expect(ll.isEmpty).toBeTruthy
+    }
+
+    it("should iterate over elements in both directions") {
+      val s = Seq("one", "two", "three")
+      val l = asJavaCollection(s)
+      val ll = new ArrayDeque[String](l)
+
+      val iter = ll.iterator()
+      for (i <- 0 until l.size()) {
+        expect(iter.hasNext()).toBeTruthy
+        expect(iter.next()).toEqual(s(i))
+      }
+      expect(iter.hasNext()).toBeFalsy
+
+      val diter = ll.descendingIterator()
+      for (i <- (0 until l.size()).reverse) {
+        expect(diter.hasNext()).toBeTruthy
+        expect(diter.next()).toEqual(s(i))
+      }
+      expect(diter.hasNext()).toBeFalsy
+    }
+  }
+}
+
```

Comment:
You also need to create the `DequeFactory` with `def allFactories` method and add it to the `CollectionFactory`:
```scala
object DequeFactory {
  def allFactories: Iterator[CollectionFactory] =
    Iterator(new ArrayDequeFactory)
}
```
```scala
object CollectionFactory {
  def allFactories: Iterator[CollectionFactory] =
    ListFactory.allFactories ++ SetFactory.allFactories ++ DequeFactory.allFactories
}
```
This way the `ArrayDeque` will be tested against all tests of `CollectionsTest` as well.